File: tests/test_license_key_link_js.js
// "I have a license key" in the trial banner opened nothing (2026-09-26).
//
// The link calls shmShowLicense(), which only switched the self-host
// modal's step to 'license'. That is enough from INSIDE the modal, where
// the overlay is already up, but the banner fires it with the overlay
// still display:none — the step flipped behind a hidden parent and the
// link read as dead. shmShowLicense must raise the overlay itself.
//
// Runs under `node tests/test_license_key_link_js.js` — no jsdom, ~50ms.
// Loads the shipped onboarding.js against a stub DOM so it tests source.

const fs = require('fs');
const path = require('path');
const vm = require('vm');

const SRC = path.join(__dirname, '..', 'clawmetry', 'static', 'js', 'onboarding.js');

let passed = 0;
let failed = 0;

function eq(actual, expected, label) {
  if (actual === expected) {
    passed++;
    console.log('  ok   ' + label);
  } else {
    failed++;
    console.log('  FAIL ' + label);
    console.log('       expected: ' + JSON.stringify(expected));
    console.log('       actual:   ' + JSON.stringify(actual));
  }
}

// The ids onboarding.js reaches for, as the dashboard serves them: the
// overlay starts hidden, every step starts hidden.
function makeDom() {
  const els = {};
  function el(id, display) {
    els[id] = { id: id, style: { display: display }, textContent: '',
                focus: function () { els[id].focused = true; } };
    return els[id];
  }
  el('selfhost-modal-overlay', 'none');
  ['home', 'otp', 'wait', 'license', 'ended'].forEach(function (s) {
    el('shm-step-' + s, 'none');
  });
  ['shm-home-error', 'shm-otp-error', 'shm-wait-error', 'shm-license-error']
    .forEach(function (id) { el(id, ''); });
  el('shm-tagline', '');
  el('shm-license-input', '');
  return els;
}

function load() {
  const els = makeDom();
  const sandbox = {
    document: {
      readyState: 'complete',
      getElementById: function (id) { return els[id] || null; },
      addEventListener: function () {},
    },
    setTimeout: function (fn) { fn(); return 0; },
    clearTimeout: function () {},
    setInterval: function () { return 0; },
    clearInterval: function () {},
    fetch: function () { return Promise.resolve({ json: function () { return {}; } }); },
    // CLOUD_MODE short-circuits _boot(): this suite is about the shm*
    // globals, not the first-run gate.
    CLOUD_MODE: true,
  };
  sandbox.window = sandbox;
  vm.createContext(sandbox);
  vm.runInContext(fs.readFileSync(SRC, 'utf8'), sandbox);
  return { win: sandbox, els: els };
}

console.log('shmShowLicense — reachable from the banner, not just from inside the modal');

const cold = load();
eq(typeof cold.win.shmShowLicense, 'function', 'shmShowLicense is exported on window');
eq(cold.els['selfhost-modal-overlay'].style.display, 'none', 'overlay starts hidden');

cold.win.shmShowLicense();
eq(cold.els['selfhost-modal-overlay'].style.display, 'flex',
   'a cold call raises the overlay (this is the bug: it stayed none)');
eq(cold.els['shm-step-license'].style.display, 'block', 'the paste step is the visible step');
eq(cold.els['shm-step-home'].style.display, 'none', 'the sign-in step is hidden');
eq(cold.els['shm-license-input'].focused, true, 'the key field takes focus');

// From inside an already-open modal the behaviour is unchanged.
const warm = load();
warm.win.openSelfhostModal();
eq(warm.els['shm-step-home'].style.display, 'block', 'openSelfhostModal still lands on home');
warm.win.shmShowLicense();
eq(warm.els['selfhost-modal-overlay'].style.display, 'flex', 'overlay stays up');
eq(warm.els['shm-step-license'].style.display, 'block', 'in-modal link still switches step');

console.log('');
console.log(failed === 0 ? 'PASS (' + passed + ')' : 'FAIL (' + failed + ' of ' + (passed + failed) + ')');
process.exit(failed === 0 ? 0 : 1);

File: tests/test_license_key_link_js.py
"""The trial banner's "I have a license key" link opened nothing (2026-09-26).

Field report from a live dashboard on an active trial ("Your trial ends in
3 days"): clicking the link next to the Get-a-license CTA did nothing at
all — no modal, no error, no console noise.

The link calls ``shmShowLicense()``, which switched the self-host modal to
its 'license' step and stopped there. Called from inside the modal — its
only caller until the banner shipped — that is correct, because
``openSelfhostModal()`` has already set ``#selfhost-modal-overlay`` to
``display:flex``. The banner calls it with the overlay still ``none``, so
the step flipped behind a hidden parent and nothing painted.

The behaviour is pinned in ``test_license_key_link_js.js``, which loads the
shipped ``onboarding.js`` against a stub DOM and asserts a cold call raises
the overlay. This file also keeps the markup side honest: the banner link
must keep calling ``shmShowLicense``, and the modal must keep shipping the
paste step it opens onto.
"""

from __future__ import annotations

import os
import shutil
import subprocess

import pytest

_HERE = os.path.dirname(os.path.abspath(__file__))
_JS_TEST = os.path.join(_HERE, "test_license_key_link_js.js")
_ROOT = os.path.dirname(_HERE)


def _read(relpath: str) -> str:
    with open(os.path.join(_ROOT, relpath), encoding="utf-8") as fh:
        return fh.read()


@pytest.mark.skipif(
    shutil.which("node") is None,
    reason="node not on PATH; JS unit tests only run when Node is available",
)
def test_license_link_unit_suite() -> None:
    proc = subprocess.run(
        ["node", _JS_TEST], capture_output=True, text=True, timeout=30
    )
    output = (proc.stdout or "") + (proc.stderr or "")
    assert proc.returncode == 0, "license-key link tests failed:\n" + output
    assert "PASS" in output, "no PASS line in output:\n" + output


def test_show_license_raises_the_overlay_itself() -> None:
    """Pins the seam: the caller must not have to open the modal first."""
    js = _read("clawmetry/static/js/onboarding.js")
    anchor = js.find("window.shmShowLicense = function")
    assert anchor != -1, "shmShowLicense must exist in onboarding.js"
    body = js[anchor:anchor + 600]
    assert "selfhost-modal-overlay" in body, (
        "shmShowLicense must raise #selfhost-modal-overlay — a step switch "
        "inside a hidden overlay paints nothing and the link reads as dead"
    )
    assert "'flex'" in body


def test_banner_link_still_targets_the_paste_surface() -> None:
    html = _read("clawmetry/templates/partials/banners.html")
    anchor = html.find('id="license-expired-have-key"')
    assert anchor != -1, "the banner's license-key link is gone"
    assert "shmShowLicense" in html[anchor:anchor + 400]
    modal = _read("clawmetry/templates/partials/selfhost-modal.html")
    assert 'id="shm-step-license"' in modal, (
        "shmShowLicense opens onto #shm-step-license; without it the link "
        "is dead again"
    )
    assert 'id="shm-license-input"' in modal
